feat(server): invalidate module with hmr

bluwy opened the pull request:
<!-- Thank you for contributing! -->

### Description

Fix #6871

New API: `server.reloadModule` that invalidates a module with hmr, unlike `moduleGraph.invalidateModule`

<!-- Please insert your description here and provide especially info about the "what" this PR is solving -->

### Additional context

<!-- e.g. is there anything you'd like reviewers to focus on? -->
Question:
1. Would the API name be confusing?
2. Is the first parameter ok? e.g. a `string` vs a `ModuleNode`

---

### What is the purpose of this pull request? <!-- (put an "X" next to an item) -->

- [ ] Bug fix
- [x] New Feature
- [ ] Documentation update
- [ ] Other

### Before submitting the PR, please make sure you do the following

- [x] Read the [Contributing Guidelines](https://github.com/vitejs/vite/blob/main/CONTRIBUTING.md).
- [x] Read the [Pull Request Guidelines](https://github.com/vitejs/vite/blob/main/CONTRIBUTING.md#pull-request-guidelines) and follow the [Commit Convention](https://github.com/vitejs/vite/blob/main/.github/commit-convention.md).
- [x] Check that there isn't already a PR that solves the problem the same way to avoid creating a duplicate.
- [x] Provide a description in this PR that addresses **what** the PR is solving, or reference the issue that it solves (e.g. `fixes #123`).
- [x] Ideally, include relevant tests that fail without this PR but pass with it.

bluwy committed:
feat(server): invalidate module with hmr

bluwy committed:
chore: fix lint

patak-dev commented:
> Would the API name be confusing?

I think that it may be confusing that `server.invalidateModule` and `moduleGraph.invalidateModule` has the same name but they have different side effects.

> Is the first parameter ok? e.g. a string vs a ModuleNode

We should probably follow the current convention in the module graph and pass a module instead.

What about `server.reloadModule(module, ...)`? IMO it more clearly states that it will trigger a Hot Module Reload for the module, and not only invalidate it.

bluwy committed:
chore: merge main

bluwy committed:
chore: update api

bluwy commented:
`reloadModule` sounds nice! I updated to use that.

> We should probably follow the current convention in the module graph and pass a module instead.

I also updated to a `ModuleNode`. I initially didn't went with it since the API would tie to the `moduleGraph`. And to keep it consistent with `ssrLoadModule`. But I feel like `ModuleNode` would be more futureproof too so it might be fine.

bluwy committed:
chore: format

rosskevin commented:
This PR would make my i18next resources HMR desperately easy (if I have the correct understanding).  Invalidating my virtual module should cause  all the relative i18next resource to refresh.  Aside from this, it looks like a detailed mess (outside of the vite ecosystem) to invalidate old resources and load new.

patak-dev commented:
Thanks for your feedback @rosskevin, it's very helpful to know real use cases for PRs 🙏🏼
This feature will be included in Vite 3.2
